modules/kvtagger/tests/test_kvtagdb.c

```diff
@@ -0,0 +1,259 @@
+
+static void
+test_insert()
+{
+  KVTagDB *tagdb = kvtagdb_new();
+
+  _fill_tagdb(tagdb, "selector", "name", "value", 2, 5);
+  int ctr = 0;
+  assert_true(kvtagdb_number_of_tags(tagdb, "selector-0") == 5, "selector-0 should have 5 nv-pairs");
+  kvtagdb_foreach_tag(tagdb, "selector-0", _foreach_ctr, (gpointer)&ctr);  
+  assert_true(ctr == 5, "foreach should find 5 nv-pairs for selector-0");
+  
+  kvtagdb_unref(tagdb);
+}
+
+static void
+test_get_selectors()
+{
+  KVTagDB *tagdb = kvtagdb_new();
+
+  _fill_tagdb(tagdb, "selector", "name", "value", 2, 5);
+
+  GList *selectors = kvtagdb_get_selectors(tagdb);
+  GList *selector0 = g_list_find_custom(selectors, "selector-0", _g_strcmp);
+  GList *selector1 = g_list_find_custom(selectors, "selector-1", _g_strcmp);
+
+  assert_string((gchar *)selector0->data, "selector-0", "");
+  assert_string((gchar *)selector1->data, "selector-1", "");
+  
+  g_list_free(selectors);
+
+  kvtagdb_unref(tagdb);
+}
+
+typedef struct _TestNVPair
+{
+  const gchar *name;
+  const gchar *value;
+} TestNVPair;
+
+typedef struct _TestNVPairStore
+{
+  TestNVPair *pairs;
+  int ctr;
+} TestNVPairStore;
+
+static void
+_foreach_get_nvpairs(gpointer arg, const TagRecord *record)
+{
+  TestNVPairStore *store = (TestNVPairStore *)arg;
+  TestNVPair pair = {.name=record->name, .value=record->value};
+  store->pairs[store->ctr++] = pair;
+}
+
+static void
+_assert_kvtag_db_contains_name_value_pairs_by_selector(KVTagDB *tagdb, 
+                                                       const gchar *selector,
+                                                       TestNVPair *expected_nvpairs,
+                                                       guint number_of_expected_nvpairs)
+{
+  TestNVPair result[number_of_expected_nvpairs];
+  TestNVPairStore result_store = {.pairs = result, .ctr = 0};
+  
+  kvtagdb_foreach_tag(tagdb, selector, _foreach_get_nvpairs, (gpointer) &result_store);
+  assert_true(result_store.ctr == number_of_expected_nvpairs, "");
+  guint i;
+  for (i = 0; i < number_of_expected_nvpairs; i++)
+    {
+      assert_string(result[i].name, expected_nvpairs[i].name, "");
+      assert_string(result[i].value, expected_nvpairs[i].value, "");
+    }
+}
+
+static void
+test_inserted_nv_pairs()
+{
+  KVTagDB *tagdb = kvtagdb_new();
+  _fill_tagdb(tagdb, "selector", "name", "value", 1, 3);
+
+ TestNVPair expected_nvpairs[3] = {
+    {.name="name-0.0", .value="value-0.0"},
+    {.name="name-0.1", .value="value-0.1"},
+    {.name="name-0.2", .value="value-0.2"}
+  };
+
+  _assert_kvtag_db_contains_name_value_pairs_by_selector(tagdb, "selector-0", expected_nvpairs, 3);
+  kvtagdb_unref(tagdb);
+}
+
+static void
+test_import_with_valid_csv()
+{
+  gchar csv_content[] = "selector1,name1,value1\n"\
+                        "selector1,name1.1,value1.1\n"\
+                        "selector2,name2,value2\n"\
+                        "selector3,name3,value3";
+  FILE *fp = fmemopen(csv_content, strlen(csv_content) + 1, "r");
```

Comment:
Note that `fmemopen()` is POSIX.1-2008, which may or may not cause platform specific issues.